Evaluate visualizations in parallel using reactive observers

hubertp opened the pull request:
### Pull Request Description
WIP 

This change adds support for executing visualizations in parallel to regular program execution.
Previously, attaching new visualizations would have to wait for the existing evaluation to finish and a visualization would be evaluated in a callback during re-execution. In the presence of expensive (long) computations the delay would be unacceptable.

Closes #10525.

### Important Notes

Enabling parallel execution created a few problems, as previously execution was always done sequentially:
- need to increase `GUEST_PARALLELISM`
- fix lack of thread-safety in `HostClassLoader`
- TBC...

TODO:
- [ ] visualizations should be updated when a `RecomputeContextRequest` is being made (i.e. all values are recomputed)
- [ ] allow for attaching visualizations to any expression, not only the cached one
- [ ] translate one-off expressions to a new approach
- [ ] modify visualization should correctly replace an existing visualization
- [ ] send failed visualizations on Panic
### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [ ] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [ ] Unit tests have been written where possible.
- [ ] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

hubertp committed:
Minor refactorings

hubertp committed:
Add rxjava dep

hubertp committed:
Fix bug in sync expressions/visualizations

hubertp committed:
Modify ExecutionService API to return futures

Most of `ExecutionService` now returns Futures that can be easier to
compose and execute via threadpools.

hubertp committed:
Implement regular visualizations via observables

hubertp committed:
Increase parallelism

hubertp committed:
HostClassLoader was not thread-safe

hubertp committed:
Fixed (again) potential deadlock in HostClassLoader

Plus some minor tweaks for random failures.

hubertp commented:
Still fighting with the consequences of enabling parallel execution.
a) If I ensure that visualization expressions are evaluated [under a write compilation lock](https://github.com/enso-org/enso/pull/13219/files#diff-3dbde2d45b1e4c5cf9dabfd23c1e7d89896ccac7ee7b8b694d65c1d6fd950037R78), then I occasionally get random deadlocks related to https://github.com/enso-org/enso/pull/12500#issuecomment-2745035967: 
[deadlock.txt](https://github.com/user-attachments/files/20610925/deadlock.txt)
b) If we feel lucky and omit that lock then, as one could expect, we start to get random/invalid failures like
```
org.enso.interpreter.service.error.MethodNotFoundException: Object Widgets does not define method get_widget_json in module Standard.Visualization.Widgets.
	at org.enso.runtime.instrument.common/org.enso.interpreter.service.ExecutionService.lambda$prepareFunctionCall$0(ExecutionService.java:128)
	at org.enso.runtime/org.enso.interpreter.runtime.ThreadManager.lambda$submit$2(ThreadManager.java:76)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1814)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:545)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:328)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:309)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1095)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:619)
	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotThreadTask$ThreadSpawnRootNode.executeImpl(PolyglotThreadTask.java:144)
	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotThreadTask$ThreadSpawnRootNode.execute(PolyglotThreadTask.java:135)
	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.executeRootNode(OptimizedCallTarget.java:819)
	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.profiledPERoot(OptimizedCallTarget.java:743)
	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.callBoundary(OptimizedCallTarget.java:667)
	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.doInvoke(OptimizedCallTarget.java:651)
	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.callDirect(OptimizedCallTarget.java:599)
	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.call(OptimizedCallTarget.java:562)
	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotThreadTask.run(PolyglotThreadTask.java:118)
	at java.base/java.lang.Thread.run(Thread.java:1447)
```

So we have to ensure that compilations done by visualizations are properly locked but at the same time do not block interpreter.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/HostClassLoader.java:
```diff
@@ -31,7 +31,13 @@ final class HostClassLoader extends URLClassLoader implements AutoCloseable {
-  public HostClassLoader() {
+  private static final HostClassLoader INSTANCE = new HostClassLoader();
+
+  public static HostClassLoader getClassLoader() {
```
- there are various attempts to rewrite this `HostClassLoader`
- one of them is #10714
- it attempts to provide [independent classloader for each library](https://github.com/enso-org/enso/pull/10714/files#r2131518443)
- making `HostClassLoader` a singleton goes _against such efforts_
- but it may be good for purposes of this PR 
- I guess we can unsingletonize the classloader later

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/HostClassLoader.java:
```diff
@@ -60,26 +66,33 @@ protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundE
-      l = findClass(name);
-      if (resolve) {
-        l.getMethods();
+    synchronized (INSTANCE) {
```
The future of `HostClassLoader` is to have multiple instances, thus this should probably be
```suggestion
    synchronized (this) {
```
otherwise we have problems with this code when removing the singleton nature of the class loader.

hubertp replied:
I believe I had before and got a deadlock with one thread keeping a lock on ConcurrentHashMap and the other having a lock on `this` both unable to proceed.

JaroslavTulach replied:
- Right now `this == INSTANCE` - because `HostClassLoader` is made singleton. E.g. there is going to be no change in behavior.
- Re. "lock on `ConcurrentHashMap`" - who is using locks on `ConcurrentHashMap`? It is supposed to be lock free...

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/EnsoContext.java:
```diff
@@ -80,7 +80,7 @@ public final class EnsoContext {
   private final EnsoLanguage language;
   private final Env environment;
-  private final HostClassLoader hostClassLoader = new HostClassLoader();
+  private final HostClassLoader hostClassLoader = HostClassLoader.getClassLoader();
```
- #10714 is also changing this line, btw.
- there are the [future plans](https://github.com/enso-org/enso/pull/13219#discussion_r2131520421)

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/HostClassLoader.java:
```diff
@@ -60,26 +66,33 @@ protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundE
+        logger.trace("Class {} found in cache", name);
+        return l;
+      }
+      if (!isRuntimeModInBootLayer && name.startsWith("org.graalvm")) {
+        return polyglotClassLoader.loadClass(name);
+      }
+      if (name.startsWith("org.slf4j")) {
+        // Delegating to system class loader ensures that log classes are not loaded again
+        // and do not require special setup. In other words, it is using log configuration that
+        // has been setup by the runner that started the process. See #11641.
+        return polyglotClassLoader.loadClass(name);
+      }
```
- What can be the source of concurrency problems? I am asking myself
- probably just the read and write to `loadedClasses`, right?
- This can be **moved out** of the `synchronized` block then

hubertp replied:
The failures were rather non-obvious but related to classloading such as
```
Caused by: java.lang.LinkageError: loader org.enso.interpreter.runtime.HostClassLoader @7dc6887e attempted duplicate class definition for org.enso.base.Text_Utils. (org.enso.base.Text_Utils is in unnamed module of loader org.enso.interpreter.runtime.HostClassLoader @7dc6887e, parent loader 'app')
```
once I synchronized this they all stopped and I've never seen them again.

JaroslavTulach replied:
- OK, thanks for sharing the failure: `attempted duplicate class definition for org.enso.base.Text_Utils`
- As far I can interpret it, it is related to two threads calling
   - into `URLClassLoader.findClass` for the same `name`
   - which then calls `defineClass` twice
   - and that throws the exception on the second attempt
- good to know the `synchronized` block fixes the problem
- it is just _too wide_ synchronization ...
- ...using the [one with Futures](https://github.com/enso-org/enso/pull/13219#discussion_r2131542150) would be more fine grain
- but such a change can certainly wait for now - it'd be good to have it before integration however

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/HostClassLoader.java:
```diff
@@ -60,26 +66,33 @@ protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundE
+          l.getMethods();
+        }
+        logger.trace("Class {} found, putting in cache", name);
+        loadedClasses.put(name, l);
```
- if the concurrency problems are related to read/write to `loadClasses`
- then we should eliminate them on the _scope of_ `name`
- e.g. don't `synchronize` to avoid parallel classloading, but
- make sure we call `findClass(name)` only once per each `name`
- I guess a future would do:
```java
    private ConcurrentHashMap<String, Future<Class<?>>> loadedClasses;

    public Class<?> loadClass(String name) throws Exception {
        var toResolve = new CompletableFuture<?>[1];
        var f = loadedClasses.computeIfAbsent(name, (t) -> {
            var newF = new CompletableFuture<Class<?>>();
            toResolve[0] = newF;
            return newF;
        });
        if (toResolve[0] != null) {
            toResolve[0].complete(findClass(name));
        }
        return f.get();
    }
```
- if multiple threads request the same `name`
   - we select one of them
   - to handle the `findClass`
   - others wait for the future to be completed 
- threads only wait if there is another one loading `name` at the same time
- different values of `name` don't influence each other

JaroslavTulach reviewed (commented):
Some thoughts on classloading.

JaroslavTulach commented on engine/runtime-instrument-common/src/main/java/org/enso/interpreter/service/ExecutionService.java:
```diff
@@ -112,21 +113,25 @@ public EnsoContext getContext() {
   }
 
-  public FunctionCallInstrumentationNode.FunctionCall prepareFunctionCall(
+  public CompletableFuture<FunctionCallInstrumentationNode.FunctionCall> prepareFunctionCall(
```
- changing the methods of `ExecutionService` to not block is proper move
  - thanks for doing that
- moreover we need to return a _"chainable"_ object - so `ExecutionService` API users can act when the computation is finished
- using `CompletableFuture` for that purpose has its downsides
   - it mixes _provider vs. user_ API
   - as it exposes `completed` & co. methods
   - we certainly do not want people who obtain the `CompletableFuture` from this method to call `completed` on it!
- can we return [CompletableStage](https://docs.oracle.com/en/java/javase/24/docs/api/java.base/java/util/concurrent/CompletionStage.html) here?
- `CompletableStage` has only the methods that we want clients of the API to use...

JaroslavTulach commented on engine/runtime-instrument-common/src/main/java/org/enso/interpreter/service/ExecutionService.java:
```diff
@@ -237,7 +234,7 @@ public void execute(
    * @param onExecutedVisualizationCallback the consumer of an executed visualization result.
    */
-  public void execute(
+  public CompletableFuture<Object> execute(
```
- the diffs are huge, but I only see this `CompletableFuture<...>` return type change
- that OK
- consider adding `@return` javadoc line

JaroslavTulach commented on engine/runtime-instrument-common/src/main/java/org/enso/interpreter/service/ExecutionService.java:
```diff
@@ -590,11 +574,11 @@ static <E extends Throwable> E raise(Class<E> type, Throwable ex) throws E {
   }
 
-  private static <T> T resultOf(Future<T> future) {
+  public static <T> T resultOf(Future<T> future) {
```
- this method doesn't belong into `ExecutionService` API
- it was fine as an internal helper method, but it is _"too ugly"_ to be exposed in the API
- moreover it is no longer used in `ExecutionService`
- please remove it from this class

JaroslavTulach commented on engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/job/ProgramExecutionSupport.scala:
```diff
@@ -209,6 +209,7 @@ object ProgramExecutionSupport {
           onCachedValueCallback,
           onExecutedVisualizationCallback
         )
+        ExecutionService.resultOf(pending)
```
- this line would be slightly problematic with pure `completionStage`
- but one can always write `stage.toCompletableFuture().get()`...

JaroslavTulach commented on engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/WarningPreview.scala:
```diff
@@ -17,15 +18,20 @@ object WarningPreview {
+        )
+      )
+    val visualizationResult =
+      ExecutionService.resultOf(visualizationResultFuture)
```
- do we have to wait here for the result?
- can't this method return `CompletableStage<String>`?
- when we have reactive `ExecutionService` - it'd be better/sane/consistent to propagate reactivity further

JaroslavTulach reviewed (commented):
- changing signatures of `ExecutionService` to return future like objects is good
- I would feel safer if we returned `CompletionStage` instead of `CompletableFuture`
  - it more clearly separates the API we want clients of `ExecutionService` to use
  - from the internal methods only for those who create `CompletableFuture`s
- please try to change the methods to return `CompletionStage`
- if it doesn't work, we can live with `CompletableFuture`, but...

JaroslavTulach commented on engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/ExecutionContextManager.scala:
```diff
@@ -188,13 +188,15 @@ class ExecutionContextManager {
+    }
   }
 
   /** Removes a visualization from the holder.
```
I've just tried to `runProjectManagerDistribution` and open a simple project. I see exceptions like:
```
[WARN] [2025-06-06T11:08:31.794] [org.enso.persist] Adding at 1269246 object:
  org.enso.compiler.core.ir.Name$Qualified@798b9e67
but there already is:
  org.enso.compiler.core.ir.Name$Qualified@724ab5aa
are they equal: true
java.io.IOException: Adding at 1269246 object:
  org.enso.compiler.core.ir.Name$Qualified@798b9e67
but there already is:
  org.enso.compiler.core.ir.Name$Qualified@724ab5aa
are they equal: true
        at org.enso.persistance/org.enso.persist.PerInputImpl.readIndirect(PerInputImpl.java:228)
        at org.enso.persistance/org.enso.persist.PerInputImpl.readObject(PerInputImpl.java:73)
        at org.enso.runtime.parser/org.enso.compiler.core.ir.IrPersistance$PersistScalaOption.readObject(IrPersistance.java:153)
        at org.enso.runtime.parser/org.enso.compiler.core.ir.IrPersistance$PersistScalaOption.readObject(IrPersistance.java:140)
        at org.enso.persistance/org.enso.persist.Persistance.readWith(Persistance.java:165)
        at org.enso.persistance/org.enso.persist.PerInputImpl.readInline(PerInputImpl.java:66)
        at org.enso.runtime.parser/org.enso.compiler.core.ir.PersistName_MethodReference.readObject(PersistName_MethodReference.java:11)
        at org.enso.runtime.parser/org.enso.compiler.core.ir.PersistName_MethodReference.readObject(PersistName_MethodReference.java:4)
        at org.enso.persistance/org.enso.persist.Persistance.readWith(Persistance.java:165)
        at org.enso.persistance/org.enso.persist.PerInputImpl.readInline(PerInputImpl.java:66)
        at org.enso.runtime.parser/org.enso.compiler.core.ir.PersistMethod_Explicit.readObject(PersistMethod_Explicit.java:11)
        at org.enso.runtime.parser/org.enso.compiler.core.ir.PersistMethod_Explicit.readObject(PersistMethod_Explicit.java:4)
        at org.enso.persistance/org.enso.persist.Persistance.readWith(Persistance.java:165)
        at org.enso.persistance/org.enso.persist.PerInputImpl.readIndirect(PerInputImpl.java:217)
        at org.enso.persistance/org.enso.persist.PerInputImpl.readObject(PerInputImpl.java:73)
        at org.enso.runtime.parser/org.enso.compiler.core.ir.IrPersistance$PersistScalaList.readObject(IrPersistance.java:240)
        at org.enso.runtime.parser/org.enso.compiler.core.ir.IrPersistance$PersistScalaList.readObject(IrPersistance.java:220)
        at org.enso.persistance/org.enso.persist.Persistance.readWith(Persistance.java:165)
        at org.enso.persistance/org.enso.persist.PerInputImpl.readInline(PerInputImpl.java:66)
        at org.enso.runtime.parser/org.enso.compiler.core.ir.PersistModule.readObject(PersistModule.java:13)
        at org.enso.runtime.parser/org.enso.compiler.core.ir.PersistModule.readObject(PersistModule.java:4)
        at org.enso.persistance/org.enso.persist.Persistance.readWith(Persistance.java:165)
        at org.enso.persistance/org.enso.persist.PerInputImpl.readInline(PerInputImpl.java:66)
        at org.enso.persistance/org.enso.persist.PerBufferReference.readObject(PerBufferReference.java:48)
        at org.enso.persistance/org.enso.persist.Persistance$Reference.get(Persistance.java:305)
        at org.enso.runtime.parser/org.enso.compiler.core.ir.IrLazyMap$En.getValue(IrLazyMap.java:57)
        at org.enso.runtime.parser/org.enso.compiler.core.ir.IrLazyMap.get(IrLazyMap.java:37)
        at org.enso.runtime/org.enso.interpreter.caches.ImportExportCache$MapToBindings.findForModule(ImportExportCache.java:145)
```
they are harmless (as there is the `are they equal: true` message and that signals everything is correct), but they clearly show `IrLazyMap` hasn't really been tested/designed for parallel access.

hubertp commented:
re propagating futures downstream:
Yeah, I've tried that a few times now but every time it becomes a bit of nightmare to compose different computations. Java is not particularly friendly in that regard. Scala would be better but I doubt that would be accepted.
Also, I want to first try to get the basic parallel execution of visualizations going, without deadlocks and other glitches, before I follow the rabbit hole.

JaroslavTulach commented on engine/language-server/src/main/scala/org/enso/languageserver/boot/MainModule.scala:
```diff
@@ -319,6 +319,10 @@ class MainModule(serverConfig: LanguageServerConfig, logLevel: Level) {
   )
+  extraOptions.put(
+    RuntimeOptions.GUEST_PARALLELISM,
+    3.toString
```
### Incremental Integration of Fixes

- I am afraid we won't be able to fix all problems in a single PR
- I would still like us to move forward
- hence I suggest to make `GUEST_PARALLELISM=3` an **opt-in** feature
- by default we would stick to the safe `GUEST_PARALLELISM=1` value
- that way we can integrate and still _give early adopters a chance to test_ the increased parallelism

### Pool of One Must be Enough

- however that means we can run with `1.toString` value
- right now, we cannot
- the code starvates for me - not sure why
- but there is no reason why we shouldn't be running with `GUEST_PARALLELISM=1`
- I'll continue investigation of this

hubertp committed:
Merge branch 'develop' into wip/hubert/10525-reactive-cache

hubertp committed:
fmt

hubertp committed:
Optional execution in PushContext request

By default, no execution is triggered by push context request.

hubertp committed:
adapt tests

hubertp committed:
More efficient visualization upsert via retries

Evaluation of a visualization occasionally requires write compilation
lock to load/compile involved modules. This leads to common lock
contention situations when in the current implementation.

Upgrading from a read compilation lock to a write one leads to deadlock
and is not supported by Reentrant locks anyway. Requesting write
compilation lock always is also not acceptable because it impacts
performance.

The current solution uses retries:
- locations that trigger compilation in visualization evaluation are
  well known
- we first attempt to evaluate visualizations without a compilation and
  read lock only
- if the latter fails, we acquire write lock and retry

hubertp committed:
fix build for tests

hubertp commented:
Most runtime visualization unit tests will now fail because of how they use cache preferences.
Unfortunately in our tests they are only set once a visualization is being upserted.
That was fine before because once upserted the triggered execution would trigger a visualization eventually.

With this PR we rely on the fact that when upserting visualizations, runtime cache has cache preferences (i.e. IDs of expressions) set correctly already. So it's a chicken and egg problem.

hubertp committed:
fix NI build

hubertp commented:
Note that with current implementation I'm no longer experiencing any deadlocks or lock contention:
[Kazam_screencast_00037.webm](https://github.com/user-attachments/assets/1407468e-0449-4530-ba93-1c7b1a348a4d)

Notice how a simple expression's visualization is loaded **before** reading of a large CSV file finishes. Previously that would be unimaginable as visualizations would only be triggered **after** the main execution is finished.

JaroslavTulach commented on build.sbt:
```diff
@@ -501,6 +501,13 @@ val circe = Seq("circe-core", "circe-generic", "circe-parser")
+
+val rxJavaVersion = "3.1.10"
+val rxJava = Seq(
+  "io.reactivex.rxjava3" % "rxjava" % "3.1.10"
```
- I am surprised by usage of `rxjava`
- sure, it stands for _Reactive Java_ - but it is _not the reactive_ I was thinking about
- when talking about reactive I mean _Vue.js reactive_ - e.g.
   - [reactive core](https://vuejs.org/api/reactivity-core.html)
   - with its `ref`, `computed`, `watchEffect`, etc.
- or about [Svelte](http://wiki.apidesign.org/wiki/Svelte)'s reactivity as shown in the following video

[![Svelte Reactivity](https://img.youtube.com/vi/vwQ06AjwKkGiUKzG/0.jpg)](https://www.youtube.com/watch?v=vwQ06AjwKkGiUKzG)

JaroslavTulach replied:
- I am surprised by usage of `rxjava`
- sure, it stands for _Reactive Java_ - but it is _not the reactive_ I was thinking about
- when talking about reactive I mean _Vue.js reactive_ - e.g.
   - [reactive core](https://vuejs.org/api/reactivity-core.html)
   - with its `ref`, `computed`, `watchEffect`, etc.
- or about [Svelte](http://wiki.apidesign.org/wiki/Svelte)'s reactivity as shown in the following video

[![Svelte Reactivity](https://img.youtube.com/vi/AdNJ3fydeao/0.jpg)](https://www.youtube.com/watch?v=AdNJ3fydeao)

JaroslavTulach commented on engine/runtime-instrument-common/src/main/java/org/enso/interpreter/instrument/RuntimeCache.java:
```diff
@@ -1,13 +1,19 @@
+import io.reactivex.rxjava3.observers.DisposableObserver;
+import io.reactivex.rxjava3.schedulers.Schedulers;
+import io.reactivex.rxjava3.subjects.ReplaySubject;
+import io.reactivex.rxjava3.subjects.Subject;
```
- once I had a discussion with Helidon guys about reactivity
- just to conclude **server side reactive is different** to client side one
   - all those _backpressures_, etc... that is not really useful on the client
- there are similarities, but I am not really knowledgeable of rxjava's `Subject`, `ReplaySubject`, etc.
- all I can tell is that I don't see `ref`, `computed`, `watchEffect` or similar
- thus bear with me...

JaroslavTulach commented on engine/runtime-instrument-common/src/main/java/org/enso/interpreter/instrument/RuntimeCache.java:
```diff
@@ -31,15 +39,78 @@ public final class RuntimeCache implements java.util.function.Function<String, O
+      Consumer<Object> onNext,
+      Consumer<Throwable> onFailure,
+      Executor executor) {
+    Subject<Object> observable;
```
- what is the role of `observable`?
- what shall it observe?
- the value associated with `UUID`?
- just the single value?

JaroslavTulach commented on engine/runtime-instrument-common/src/main/java/org/enso/interpreter/instrument/RuntimeCache.java:
```diff
@@ -31,15 +39,78 @@ public final class RuntimeCache implements java.util.function.Function<String, O
+          observables.computeIfAbsent(expressionId, (UUID uuid) -> ReplaySubject.createWithSize(1));
+    }
+    var scheduler = Schedulers.from(executor);
+    var observer =
```
- am I wrong to say that this is just like `CompletableFuture` that others may "chain" actions to
- which is triggered when the value is changed?
- for something as simple we probably don't need `rxjava` 3rd party library
- there are methods like `thenApply(..., executor)` already in the JDK...
- but when talking about _"reactive cache"_ I probably wanted something completely different

JaroslavTulach reviewed (commented):
# Reactive as in Vue.js

- This is a comment about _reactive_ `RuntimeCache`. 
- For few days I was puzzled why `rxjava`?
- Just to realize that the _RxJava reactivity_ isn't the reactivity me and @Frizi were talking about in Italy

### Motivation: `Table.join`

- there are _"visualizations"_ that no longer depend just on a single UUID value
- for example `Table.join` depends on `self` and `other`
- support for that was implemented in #9410
- but the support is fragile
- what we want is `watchEffect`...
   - visualization may access as many UUID values from the cache as it wants
   - the **execution is tracked** and all the accessed UUIDs are recorded
   - once any of the UUIDs "observables" changes
   - re-evaluation of the visualization is triggered
   - the **execution is tracked** and all the accessed UUIDs are recorded
   - and so on, so on
- these kinds of visualizations that observer many values from the cache will be more and more common
- especially if we connect Y.doc structures with the cache

### Motivation: Eviction from the Cache

- when the IDE modifies a line of code that defines a variable
- it is clear the value associated with that variable needs to be removed from the cache
- and recomputed on next computation
- however _not only the single value_ needs to be evicted
- all other that depend on it need to be evicted as well
- right now we complicate and slow our compiler down with passes like `DataflowAnalysis` & co.
-  instead - we should use `watchEffect`
   - when evaluating a component/node/assignment to a variable
   - **track the execution** and record all UUIDs that are accessed while computing the value
   - when putting the `computed` value into the runtime cache
   - let `watchEffect` connect it with all the observable UUIDs it read
   - as soon as one of them is evicted, evict also all the dependent ones
- this is the Vue.js, Knockout.js & co. **client reactivity** 
- that would allow us to do less during static analysis
- and get better results (as observing runtime has smaller closure than static analysis)

## Java Implementation

- I don't think rxjava is a suitable library for this kind of _client reactivity_
- there is [HTML/Java API](https://bits.netbeans.org/html+java/dev/) that has the right "observables" we need

hubertp committed:
drop unused parameter

hubertp committed:
fix typo

hubertp committed:
Make sure observers/visualizations can be modified

hubertp committed:
Infer diagnostics in Truffle context

When diagnostics are being gathered outside of ThreadManager, execution
will blow up due to Truffle restrictions.

hubertp committed:
Ensure visualization upsert triggers execution when necessary

`ExecuteJob` needs to run whenever there was no observable present for
the requested expression ID. Had to bring back cache invalidation logic
to ensure that one can attach visualization to any expression. This was
problematic for situations when requested expression was part of a
larger one which result has already been cached.

hubertp committed:
Adapt tests to new implementation

hubertp committed:
Merge branch 'develop' into wip/hubert/10525-reactive-cache

hubertp committed:
Fix remaining tests

A bug lurked in `HostClassLoader` that was manifesting itself in a
hidden way (no symbol found). Now logging the exception so that it is
easier to spot the problem.

hubertp committed:
Reply to `AttachVisualization` request immediately

Bringing back previous behaviour. Otherwise, it is very likely request
handlers will start reporting timeouts - evaluation/execution of
visualizations may make a while.

hubertp committed:
Evaluate vis expression only once

No need to recalculate expression every time.

hubertp committed:
Partial revert of evaluate-once change

When evaluating module only once, modifications to visualization modules
will not be picked up, as indicated by a failing test.

hubertp committed:
Workaround notorious starvation issues in Truffle

Parked threads conflict with guest truffle threads that are waiting at
synchronization points. To minimize chances of thread starvation we
ensure that core threads are not kept alive when done.

hubertp commented:
Obsolete via #13948
